enhance(backend): 通知がミュート・凍結を考慮するようにする

tai-cha opened the pull request:
<!-- ℹ お読みください / README
PRありがとうございます！ PRを作成する前に、コントリビューションガイドをご確認ください:
Thank you for your PR! Before creating a PR, please check the contribution guide:
https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md
-->

## What
<!-- このPRで何をしたのか？ どう変わるのか？ -->
<!-- What did you do with this PR? How will it change things? -->

- ミュートしたユーザー・凍結したユーザーからの通知が出ないようにした

- 副産物としてnotifierIdが指定されているのにUserが存在しない場合にも通知が弾かれるようになった

- #13284 と同時にマージしたいっぽいので #13284 をマージした

- refactor: packとpackGrouped, packManyとpackManyGroupedがあまりに共通すぎたのでもう共通化した

## Why
<!-- なぜそうするのか？ どういう意図なのか？ 何が困っているのか？ -->
<!-- Why do you do it? What are your intentions? What is the problem? -->

resolve #13411 
resolve #13316

#13284 より
Fix https://github.com/misskey-dev/misskey/issues/10650
Fix https://github.com/misskey-dev/misskey/issues/13369

## Additional info (optional)
<!-- テスト観点など -->
<!-- Test perspective, etc -->

close #13284

## Checklist
- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [ ] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [x] (If needed) Update CHANGELOG.md
- [x] (If possible) Add tests

dakkar committed:
Never return broken notifications #409

Since notifications are stored in Redis, we can't expect relational
integrity: deleting a user will *not* delete notifications that
mention it.

But if we return notifications with missing bits (a `follow` without a
`user`, for example), the frontend will get very confused and throw an
exception while trying to render them.

This change makes sure we never expose those broken notifications. For
uniformity, I've applied the same logic to notes and roles mentioned
in notifications, even if nobody reported breakage in those cases.

Tested by creating a few types of notifications with a `notifierId`,
then deleting their user.

(cherry picked from commit 421f8d49e5d7a8dc3a798cc54716c767df8be3cb)

kakkokari-gtyih committed:
Update Changelog

kakkokari-gtyih committed:
Merge branch 'develop' into fix-10650

kakkokari-gtyih committed:
Merge branch 'develop' into fix-10650

kakkokari-gtyih committed:
Merge branch 'develop' into fix-10650

kakkokari-gtyih committed:
Update CHANGELOG.md

tai-cha committed:
enhance: 通知がミュートを考慮するようにする

tamaina commented:
凍結も…

tai-cha commented:
別トピックだから同時にやらないほうがいいかなと思ったんですが凍結も一緒にしちゃっていいか

tamaina commented:
マージするのが大変なんじゃないかしら

tamaina commented:
（コンフリクトしなくても個人的には一緒でいいと思っている（PRの粒度については諸説ある

tai-cha committed:
enhance: 通知が凍結も考慮するようにする

tai-cha commented:
テスト書いたらdraft外したい

tamaina commented:
（はっや（ここまで簡単だと思ってなかった

tai-cha commented:
そりゃUserの実体取らざるを得ないから凍結も取得簡単

tamaina commented:
notifierIdを持たないタイプの通知がnullになって全部消されてる気がする

tai-cha commented:
あ！（なおします）

tamaina commented:
あとこのロジックをNotificationEntityService.pack系に書かなかったのはなぜかしら

tai-cha committed:
fix: notifierIdがない通知が消えてしまう問題

tai-cha commented:
> あとこのロジックをNotificationEntityService.pack系に書かなかったのはなぜかしら

- とりあえず書いたから

そのほうが良さそうなのでテスト書いた後にやるか

tai-cha commented:
ついでに言うとGroupedとそうじゃないもので二重になってる箇所があるのでリファクタしたい

tai-cha committed:
Add tests (通知がミュートを考慮しているかどうか)

tai-cha committed:
fix: notifierIdがない通知が消えてしまう問題 (grouped)

tai-cha commented on packages/backend/src/server/api/endpoints/i/notifications-grouped.ts:
```diff
@@ -6,7 +6,8 @@
 import * as Redis from 'ioredis';
 import { Inject, Injectable } from '@nestjs/common';
-import type { NotesRepository } from '@/models/_.js';
+import { not } from 'ajv/dist/compile/codegen/index.js';
```
なにこれ
```suggestion
```

tamaina replied:
noteとか書こうとしてnotの時点でenterしてVS Code(かなんかのIDE)のサジェスチョンが突っ込まれたと予想

tai-cha committed:
Remove unused import

tamaina commented:
>  enhance: 通知がミュート・通知を考慮するようにする

おちつけ

tai-cha commented:
凍結って打ったつもりしかなかった

tai-cha committed:
Fix: typo

tai-cha committed:
Merge branch 'develop' into notification-hide-muting-user

tai-cha commented:
（Serviceが要求を返せない場合のベストプラクティスがわかっていない）

tai-cha committed:
Revert "enhance: 通知が凍結も考慮するようにする"

This reverts commit b1e57e571dfd9a7d8b2430294473c2053cc3ea33.

tai-cha committed:
Revert API handling

tai-cha committed:
Remove unused imports

tai-cha committed:
enhance: Check if notifierId is valid in NotificationEntityService

tai-cha commented:
通知の発行時の挙動さえ決めちゃえば多分完成

tai-cha commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -64,8 +67,11 @@ export class NotificationEntityService implements OnModuleInit {
 			packedUsers: Map<MiUser['id'], Packed<'UserLite'>>;
 		},
-	): Promise<Packed<'Notification'>> {
+	): Promise<Packed<'Notification'>|null> {
 		const notification = src;
+
+		if (!(await this.#isValidNotifier(notification, meId))) return null;
+
```
ここ

tai-cha commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -159,8 +167,11 @@ export class NotificationEntityService implements OnModuleInit {
 			packedUsers: Map<MiUser['id'], Packed<'UserLite'>>;
 		},
-	): Promise<Packed<'Notification'>> {
+	): Promise<Packed<'Notification'>|null> {
 		const notification = src;
+
+		if (!(await this.#isValidNotifier(notification, meId))) return null;
+
```
ここ

tai-cha reviewed (commented):
nullでいいんかな？のやつ

tai-cha commented:
別に作成時ぐらいにしか使われないpackは今回はいじらない手もあるか？と思ったけど通知の作成イベントが飛んでしまうな

tai-cha commented:
（なんで今不完全なのにテスト全部通ってるんだ・・・）

tai-cha commented:
あー別にinternalであるから帰ってきてる結果がなくても今のところ問題がないんだ

tai-cha committed:
通知作成時にpackしてnullになったらあとの処理をやめる

tai-cha commented:
あーpackManyって結局pack呼ぶのか（勘違いしていた）

tai-cha committed:
Remove duplication of valid notifier check

tai-cha committed:
add filter notification is not null

tai-cha committed:
Merge branch 'develop' into notification-hide-muting-user

tamaina commented:
なんかねという感じはある

(凍結やミュートをチェックしたいときは外からはpackMany([MiNotification])みたいに呼ぶという運用にすればいいのかも

tai-cha commented:
packManyでやった方が問い合わせが少ないのでパフォーマンスは高いんだよな（オプション引数とってチェックする部分スキップさせればいいのか）

tai-cha committed:
Revert "Remove duplication of valid notifier check"

This reverts commit 239a6952f717add53d52c3e701e7362eb1987645.

tai-cha committed:
Improve performance

tai-cha committed:
Fix packGrouped

tai-cha commented:
多分めんどくさい運用考えなくてもこれで直感的に書けるでしょの気持ち（テスト通ればこれで一旦完成の気持ち）

tai-cha commented:
本当はミュートしたら通知リロードさせたいけどめんどくさい

tai-cha commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -277,9 +290,67 @@ export class NotificationEntityService implements OnModuleInit {
+			if (notifier.isSuspended) return null;
+
+			return notification;
+		}))) as [T|null] ).filter((notification): notification is T => notification !== null);
```
```ts
		}))) as [T|null] ).filter((notification): notification is T => notification !== null);
```
`[T|null]`を明示しないと`Awaited<T|null>`としか認識してくれなかったので困ったやつ

tai-cha committed:
Refactor: 判定部分を共通化

tai-cha committed:
Fix condition

tai-cha committed:
use isNotNull

tai-cha commented:
もしかして：
https://github.com/misskey-dev/misskey/pull/13284

とまるっきり競合するかも

tai-cha committed:
Update CHANGELOG.md

tai-cha committed:
Merge branch 'develop' into notification-hide-muting-user

tamaina commented:
> 本当はミュートしたら通知リロードさせたいけどめんどくさい

これは別prにしたほうがよさそう

tai-cha commented:
とりあえずこれマージされそうだったら別Issueにしよう

tai-cha committed:
Merge branch 'fix-10650' of https://github.com/kakkokari-gtyih/misskey into notification-hide-muting-user

tai-cha commented:
#13284 をここにマージした

tai-cha committed:
filterの改善

tai-cha committed:
Refactor: DONT REPEAT YOURSELF
Note: GroupedNotificationはNotificationの拡張なのでその例外だけ書けば基本的に共通の処理になり複雑な個別の処理は増えにくいと思われる

tai-cha commented:
Note: GroupedNotificationはNotificationの拡張なのでそれに対する例外だけ書けば基本的に共通の処理になり複雑な個別の処理は増えにくいと思われる

tai-cha committed:
Add groupedNotificationTypes

tai-cha committed:
Update misskey-js typedef

syuilo commented:
pack時に弾く感じかしら

tai-cha commented:
pack時に弾くようにした（さいしょはapiで弾いていたけどこっちのほうが良くないって@tamaina に言われてたしかにとなったため）

tai-cha commented:
packかpackManyどちらかで弾かれる（Manyの場合は複数一気にやる、パフォーマンス上の都合）

syuilo commented:
自分がミュートしたか、ユーザーが凍結された時にRedisから消した方がスッキリしそうなのとパフォーマンス的にも良さそうな気がした

tai-cha commented:
すでに来てるやつは消えないけど一理ある

syuilo commented:
すでに来てるやつは消えない というと？

tamaina commented:
> Redisから消した方が

ｗ読んでなさそう

tai-cha commented:
それだったらこの前まで書いてた個別削除のロジックを流用した別PRにしたほうがいいかも（このPRずっと見てた @tamaina さんはどっちがいいと思うか聞きたさはある）

tai-cha commented:
あーredisから個別で対象の通知消そうとすると消すときにたくさんnotificationsを読む必要があるからいうほどパフォーマンス良くないかも

syuilo commented:
ミュートや凍結が行われる数 < 通知が読まれる数
と思う

tai-cha commented:
スパムとかたくさん湧くと結構凍結発生するのとサーバーブロックの場合はたくさんミュートすることになるかも、流石に通知を読む数のが少ないけど

tamaina commented:
Redisから消すの、notifierIdをRedisで絞って消せるならいいけどそうじゃないならRedisの全部の通知をJSON.parseして処理することになりそう？

tai-cha commented:
> あーredisから個別で対象の通知消そうとすると消すときにたくさんnotificationsを読む必要があるからいうほどパフォーマンス良くないかも

これ、凍結の場合は全ユーザー分の探索が必要なのでやばい

syuilo commented:
データの持ち方を工夫すれば大丈夫そう

tai-cha commented:
どう工夫すればいいか今のところアイデアがない

syuilo commented:
考える

tamaina commented:
通知また全消しするのか…とりあえずいまはpackでサクッと実装したい気分

syuilo commented:
通知また全消しするのか というと？

tai-cha commented:
> Redisから消すの、notifierIdをRedisで絞って消せるならいいけどそうじゃないならRedisの全部の通知をJSON.parseして処理することになりそう？

実際にキーを増やすとか変えるじゃないと多分そうなってしまう

tamaina commented:
> 通知また全消しするのか というと？

データの持ち方を変えるとなるとそういうこともありそう

syuilo commented:
変え方によってはそういうこともあるけど新たにインデックスのようなデータを持たせることを考えている

tai-cha commented:
消さないでやろうとするとFTTの条件のやつみたいにキーを増やしてそれを一気に取得するとかになりそうだけどそれはそれで

tamaina commented:
古いデータ消せないなと思ったけどまあ別にいいか

tai-cha commented:
まあ古いデータは全削除してもらうのがいいのかも、でも持ち方の変え方次第では全削除の方にも影響するな

tai-cha commented:
今のやり方でも毎packManyごとに重い読み込みが発生しなければいい説があるのでそっちの線の考えるか

tai-cha committed:
Merge branch 'develop' of https://github.com/misskey-dev/misskey into notification-hide-muting-user

tai-cha committed:
Refactor: less sql calls

tamaina commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -269,17 +210,139 @@ export class NotificationEntityService implements OnModuleInit {
+		notification: T,
+		meId: MiUser['id'],
+	) : Promise<boolean> {
+		const [
+			userIdsWhoMeMuting,
+			userMutedInstances,
+		] = await Promise.all([
+			this.cacheService.userMutingsCache.fetch(meId),
+			this.cacheService.userProfileCache.fetch(meId).then(p => new Set(p.mutedInstances)),
+		]);
+
+		const notifiers = (await Promise.all([notification]
+			.map(x => 'notifierId' in x ? x.notifierId : null)
+			.filter(isNotNull)
+			.map(async id => await this.usersRepository.findOneBy({ id }))))
+			.filter(isNotNull);
+
+		return this.#validateNotifier(notification, userIdsWhoMeMuting, userMutedInstances, notifiers);
```
```suggestion
		return (await this.#filterValidNotifier([notification], meId)).length === 1;
```

tamaina committed:
refactor

tamaina committed:
clean up

tamaina commented:
packManyでNoteReadService.readを行うことで`this.notesRepository.findBy({ id: In(noteIds) })`の回数を減らしてみた

tamaina commented:
NOTE_REQUIRED_NOTIFICATION_TYPESのチェックっているの?

tamaina committed:
filter notes to mark as read

tamaina committed:
packed noteがmapなのでそちらを使う

tamaina committed:
if (notesToRead.size > 0)

tamaina committed:
if (notes.length === 0) return;

tamaina committed:
fix

tamaina committed:
Revert "if (notes.length === 0) return;"

This reverts commit 22e2324f9633bddba50769ef838bc5ddb4564c88.

tamaina committed:
:art:

tamaina committed:
console.error

tamaina committed:
err

tamaina committed:
remove try-catch

tamaina committed:
不要なジェネリクスを除去

tai-cha commented:
寝てたらめっちゃ変わってる、確認するか

tai-cha reviewed (commented):
めっちゃ綺麗になっている上にパフォーマンス向上していそう

tai-cha committed:
Merge branch 'develop' of https://github.com/misskey-dev/misskey into notification-hide-muting-user

tai-cha commented:
> packに副作用があるのはちょっと宜しくないかも

というとpackで通知を弾くのではなくこのPRで最初に実装したようにAPIコールの方でpackの前や後に凍結やミュートを考慮したほうが良いという話？

syuilo commented:
pack内でノートの既読処理はしない方が良さそう

tai-cha commented:
あーなるほど
それ直したら他には問題なさそうですか？

tai-cha committed:
Revert  (既読処理をpack内で行うものを元に戻す)

tai-cha commented:
既読処理の部分元通りにした

tai-cha committed:
Merge branch 'develop' of https://github.com/misskey-dev/misskey into notification-hide-muting-user

tai-cha committed:
Clean

tai-cha commented:
e2e落ちるけど
  ● Account Move › Local to Local › A locked account automatically accept the follow request if it had already accepted the old account.

だからこのPRの変更と関係なさそう

tai-cha committed:
Merge branch 'develop' into notification-hide-muting-user

tai-cha committed:
Merge branch 'develop' into notification-hide-muting-user

syuilo commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -269,17 +206,115 @@ export class NotificationEntityService implements OnModuleInit {
+	/**
+	 * notifierが存在するか、ミュートされていないか、サスペンドされていないかを確認するvalidator
+	 */
+	async #validateNotifier <T extends MiNotification | MiGroupedNotification> (
```
async要素ないかも

tai-cha replied:
最初はあったから名残かも（確認して必要に応じて消します）

tai-cha replied:
Fixed

syuilo commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -52,138 +54,61 @@ export class NotificationEntityService implements OnModuleInit {
-			}));
+			}))).filter(r => isNotNull(r.user));
+			// if all users have been deleted, don't show this notification
+			if (!reactions.length) {
```
```suggestion
			if (reactions.length === 0) {
```

syuilo commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -192,14 +117,19 @@ export class NotificationEntityService implements OnModuleInit {
-			}));
+			}))).filter(isNotNull);
+			// if all users have been deleted, don't show this notification
+			if (!users.length) {
```
```suggestion
			if (users.length === 0) {
```

syuilo committed:
Update packages/backend/src/core/entities/NotificationEntityService.ts

syuilo committed:
Update packages/backend/src/core/entities/NotificationEntityService.ts

syuilo commented on packages/backend/src/core/NotificationService.ts:
```diff
@@ -163,6 +163,8 @@ export class NotificationService implements OnApplicationShutdown {
 
 		const packed = await this.notificationEntityService.pack(notification, notifieeId, {});
 
+		if (packed === null) return null;
```
```suggestion
		if (packed == null) return null;
```

syuilo commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -269,17 +206,115 @@ export class NotificationEntityService implements OnModuleInit {
+		const filteredNotifications = ((await Promise.all(notifications.map(async (notification) => {
+			const isValid = await this.#validateNotifier(notification, userIdsWhoMeMuting, userMutedInstances, notifiers);
+			return isValid ? notification : null;
+		}))) as [T|null] ).filter(isNotNull);
```
```suggestion
		}))) as [T | null] ).filter(isNotNull);
```

syuilo commented on packages/backend/src/core/entities/NotificationEntityService.ts:
```diff
@@ -269,17 +206,115 @@ export class NotificationEntityService implements OnModuleInit {
+
+		const notifier = notifiers.find(x => x.id === notification.notifierId) ?? null;
+
+		if (notifier === null) return false;
```
```suggestion
		if (notifier == null) return false;
```

tai-cha replied:
:+1::bow:

syuilo committed:
Update packages/backend/src/core/entities/NotificationEntityService.ts

syuilo committed:
Update packages/backend/src/core/entities/NotificationEntityService.ts

syuilo committed:
Update packages/backend/src/core/NotificationService.ts

tai-cha committed:
Clean

syuilo commented:
🙏
